test/buffer_test.dart

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+import 'dart:async';
+
+import 'package:test/test.dart';
+
+import 'package:stream_transform/stream_transform.dart';
+
+void main() {
+  var streamTypes = {
+    'single subscription': () => new StreamController(),
+    'broadcast': () => new StreamController.broadcast()
+  };
+  for (var triggerType in streamTypes.keys) {
+    for (var valuesType in streamTypes.keys) {
+      group('Trigger type: [$triggerType], Values type: [$valuesType]', () {
+        StreamController trigger;
+        StreamController values;
+        List emittedValues;
+        bool valuesCanceled;
+        bool triggerCanceled;
+        bool isDone;
+        List errors;
+        StreamSubscription subscription;
+
+        setUp(() async {
+          valuesCanceled = false;
+          triggerCanceled = false;
+          trigger = streamTypes[triggerType]()
+            ..onCancel = () {
+              triggerCanceled = true;
+            };
+          values = streamTypes[triggerType]()
+            ..onCancel = () {
+              valuesCanceled = true;
+            };
+          emittedValues = [];
+          errors = [];
+          isDone = false;
+          subscription = values.stream
+              .transform(buffer(trigger.stream))
+              .listen(emittedValues.add, onError: errors.add, onDone: () {
+            isDone = true;
+          });
+        });
+
+        test('does not emit before `trigger`', () async {
+          values.add(1);
+          await new Future(() {});
+          expect(emittedValues, isEmpty);
+          trigger.add(null);
+          await new Future(() {});
+          expect(emittedValues, [
+            [1]
+          ]);
+        });
+
+        test('emits immediately if trigger emits before a value', () async {
+          trigger.add(null);
+          await new Future(() {});
+          expect(emittedValues, isEmpty);
+          values.add(1);
+          await new Future(() {});
+          expect(emittedValues, [
+            [1]
+          ]);
+        });
+
+        test('groups values between trigger', () async {
+          values.add(1);
+          values.add(2);
+          await new Future(() {});
+          trigger.add(null);
+          values.add(3);
+          values.add(4);
+          await new Future(() {});
+          trigger.add(null);
+          await new Future(() {});
+          expect(emittedValues, [
+            [1, 2],
+            [3, 4]
+          ]);
+        });
+
```

Comment:
A couple other suggestions for tests:

- mixing pre-emptive triggers and normal triggers
- multiple triggers in a row in between values